latch_cli/services/cp/progress.py

```diff
@@ -0,0 +1,142 @@
+from multiprocessing import BoundedSemaphore
+from multiprocessing.managers import BaseManager
+from typing import List, Optional
+
+import tqdm
+
+
+def get_progress_bar():
+    return tqdm.tqdm(
+        total=0,
+        leave=False,
+        smoothing=0,
+        unit="B",
+        unit_scale=True,
+        unit_divisor=1000,
+    )
+
+
+class ProgressBars:
+    def __init__(
+        self,
+        num_task_bars: int,
+        *,
+        show_total_progress: bool = True,
+        verbose: bool = False,
+    ):
+        if show_total_progress:
+            self.total_bar = get_progress_bar()
+            self.total_bar.desc = "Copying Files"
+            self.total_bar.colour = "green"
+            self.total_bar.unit = ""
+        else:
+            self.total_bar = None
+
+        self.total_sema = BoundedSemaphore(1)
+
+        self.verbose = verbose
+        self.verbose_sema = BoundedSemaphore(1)
+
+        self.task_bars: List[tqdm.tqdm] = [
+            get_progress_bar() for _ in range(num_task_bars)
+        ]
+        self.free_indices = {i for i in range(num_task_bars)}
+        self.bar_sema = BoundedSemaphore(num_task_bars)
+
+        self.usage = [0] * num_task_bars
+
+    def num_bars(self) -> int:
+        return len(self.task_bars)
+
+    def write(self, msg: str):
+        if not self.verbose:
+            return
+
+        with self.verbose_sema:
+            tqdm.tqdm.write(msg)
+
+    def get_free_task_bar_index(self) -> int:
```

Comment:
this + `return_task_bar` should be wrapped by a context manager (obviously locally per process)